### PR TITLE
feat: Implement RAMF serialization

### LINKED ISSUE #1: Implement RAMF serialisation
Per RS-001: https://github.com/relaynet/specs/blob/master/rs001-ramf.md

See implementation in PoC:

- https://github.com/relaynet/poc/blob/master/core/serialization.js
- https://github.com/relaynet/poc/blob/master/core/_cms.js

### PR DESCRIPTION
**Pull request by gnarea**
I'm not very happy with the ASN.1 encoders in Java/Kotlin because I couldn't find a high-level encoder that abstracted away references to ASN.1 classes and tags, like asn1js does in JavaScript. Also, I found that most encoders generate Java code from ASN.1 modules, whilst others (like BouncyCastle) leave you to your own devices, so this felt like an exercise of picking a poison. In the end I played with [jASN1](https://www.beanit.com/asn1/) and [asn1-tool](https://github.com/yafred/asn1-tool), both of which generate Java code, and picked the former because it's more widely used.

Summary:

- Generated jASN1 serialiser code, then converted it to Kotlin and refactored it.
- Added unit tests for generated code. I used the ASN.1 library in Bouncy Castle to validate the DER serialization produced by the jASN1-generated code.
- Removed dummy `Library` class since we now have real code and tests.
- I've tested the generated DER on the [ASN.1 Playground](https://asn1.io/asn1playground/) and it matches the following schema:
  ```
  Relaynet-RAMF DEFINITIONS AUTOMATIC TAGS ::=
  BEGIN
    RAMFMessage ::= SEQUENCE
    {
       recipientAddress   VisibleString (SIZE(0..1023)),
       messageId          VisibleString (SIZE(0..255)),
       creationTimeUtc    DATE-TIME,
       ttl                INTEGER (0..15552000),
       payload            OCTET STRING (SIZE(0..8388608))
    }
  END
  ```

Fixes #1.

### CODE REVIEW

File: src/main/kotlin/tech/relaycorp/relaynet/ramf/RAMFSerializer.kt [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        output.write(concreteMessageType.toInt())
+        output.write(concreteMessageVersion.toInt())
+        output.write(serializeFields())
+
+        return output.toByteArray()
+    }
+
+    @Throws(IOException::class)
+    private fun serializeFields(): ByteArray {
+        val reverseOS = ReverseByteArrayOutputStream(1000, true)
```
Comment (gnarea) [inline src/main/kotlin/tech/relaycorp/relaynet/ramf/RAMFSerializer.kt]: The second argument specifies whether the underlying buffer should be resized as the array grows. I couldn't write a test because I didn't know how to spy on it: https://stackoverflow.com/questions/60547003/how-to-spy-on-a-class-constructor-from-an-external-dependency-in-mockk

I'll write that test if/when I get a response.

File: src/test/kotlin/tech/relaycorp/relaynet/ramf/RAMFMessageTest.kt [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            fun `Timezone may be unset`() {
+            }
+
+            @Test
+            @Disabled("Pending implementation")
+            fun `Timezone may be set to UTC`() {
+            }
+        }
+    }
+}
```
Comment (gnarea) [inline src/test/kotlin/tech/relaycorp/relaynet/ramf/RAMFMessageTest.kt]: This is technically part of #9 but I thought of those edge cases and thought it'd be good to create placeholders for them. I probably should've added a comment in #9 in hindsight.

**Comment by gnarea** [comment]
I've just unassigned this from you @extropyCoder because I noticed a bug which I might not get round to fixing today.

**Comment by gnarea** [comment]
The issue I thought I found wasn't really a problem but I did realise I forgot to validate the length of the fields, which I've now done.

File: build.gradle.kts [hunk 3]
```diff
@@ -93,7 +98,7 @@ tasks.test {
     finalizedBy("jacocoTestReport")
     doLast {
         println("View code coverage at:")
-        println("file://$buildDir/reports/jacoco/test/html/index.html")
+        println("file://$buildDir/reports/coverage/index.html")
```
[the remark this answers is not in the document]
Comment (extropyCoder) [reply build.gradle.kts]: oh i also put in a fix for that

File: src/test/kotlin/tech/relaycorp/relaynet/ramf/RAMFMessageTest.kt (hunk 2 above)
Comment (extropyCoder) [reply src/test/kotlin/tech/relaycorp/relaynet/ramf/RAMFMessageTest.kt]: the TTL you have is for 180 days rather than 90

**Comment by gnarea** [comment]
Good catch! I've just fixed it and will merge this PR once CI completes.
